[IMP] payment_xendit: tokenization

nni-odoo opened the pull request:
Currently, in the APAC region, there are no payment providers
implemented that support tokenization. This commit adds tokenization
support for the `Card` payment method through Xendit. As tokenization
can not be implemented via form redirect, `Card` payments will
specifically be using a direct flow. The changes made here will only
affect `Card` payments, while other payment methods will continue using
redirect implementation.

task-3704600

chevalierv reviewed (changes requested):
Hello @nni-odoo 

Thank you for your work on this PR! I didn't have time to finish the review today so it's only a first part :)

Can you take some time to fix the CI/style issues, and make an upgrade script to remove the view? (See the red runbot) Also, please write a good commit message and document your changes in the `readme.md` file of `payment_xendit`.

Have a nice day!

chevalierv commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,199 @@
+            return;
+        }
+        // if tokenize, create multi-use token
+        const payload = {
```
`currency` should be passed to Xendit to avoid discrepancies between the platforms. According to [Xendit's documentation](https://developers.xendit.co/api-reference/#create-token), if `currency` is not given, it defaults based on your business country.

nni-odoo commented on addons/payment_xendit/models/payment_provider.py:
```diff
@@ -45,7 +58,7 @@ def _get_default_payment_method_codes(self):
         return const.DEFAULT_PAYMENT_METHODS_CODES
 
-    def _xendit_make_request(self, payload=None):
+    def _xendit_make_request(self, endpoint, endpoint_param=None, payload=None, method='POST'):
```
was following adyen for this one, but indeed there seems to be no use here, will remove it thanks

nni-odoo commented on addons/payment_xendit/models/payment_provider.py:
```diff
@@ -45,7 +58,7 @@ def _get_default_payment_method_codes(self):
         return const.DEFAULT_PAYMENT_METHODS_CODES
 
-    def _xendit_make_request(self, payload=None):
+    def _xendit_make_request(self, endpoint, endpoint_param=None, payload=None, method='POST'):
```
yep, will remove this too

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,197 @@
+        // if tokenize, create multi-use token
+        const payload = {
+            ...this.xenditGetPaymentDetails(paymentOptionId, paymentMethodCode),
+            is_multiple_use: true,
```
`tokenization_requested` or any variable that mentions tokenization exist in processingValues, will add it from the `_get_specific_processing_values`

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,197 @@
+            is_multiple_use: true,
+            should_authenticate: false,        }
+
+        Xendit.setPublishableKey(this.xenditData[paymentOptionId]['public_key'])
```
putting it on `_prepareInlineForm` after JS is loaded

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,197 @@
+            should_authenticate: false,        }
+
+        Xendit.setPublishableKey(this.xenditData[paymentOptionId]['public_key'])
+        Xendit.card.createToken(payload, (err, token) => this._xenditHandleResponse(err, token, processingValues))
```
yep will add this flow

chevalierv replied:
nitpicking: I would have done that in `_initiatePaymentFlow` to be able to set a [custom validity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity)

nni-odoo commented:
@chevalierv thank you for the review! made some amendments per your suggestions!

chevalierv commented on addons/payment_xendit/models/payment_transaction.py:
```diff
@@ -140,11 +152,48 @@ def _process_notification_data(self, notification_data):
         if payment_status in const.PAYMENT_STATUS_MAPPING['pending']:
             self._set_pending()
         elif payment_status in const.PAYMENT_STATUS_MAPPING['done']:
+            if self.tokenize and not self.token_id:
```
Why do you need to check if a token is not already created? This situation should be impossible as we write the token and `tokenize` at the same time on the transaction. 
```suggestion
            if self.tokenize:
```

chevalierv commented on addons/payment_xendit/models/payment_transaction.py:
In the very unlikely case where `_get_specific_rendering_values` when using the `Card` payment method, an error will be thrown preventing making the operation (validation|payment). I'd suggest adding a check based on the payment method.

https://github.com/odoo/odoo/blob/e069c3a0db3dbcf986d62acac1c9aa3ed36a2a19/addons/payment_xendit/models/payment_transaction.py#L42
```diff
-     if self.provider_code != 'xendit':
+     if self.provider_code != 'xendit' and self.payment_method != 'card':
```

chevalierv reviewed (changes requested):
Hello @nni-odoo,

Here is my second review pass :duck: 

I've pushed some nitpicking changes mainly about docstrings and also, I've changed my mind about the `is_multiple_use` thing so I've reverted the changes myself :) In the end, keeping it like this is cleaner and allows us to avoid adding more things like `_get_specific_processing_values`. I've moved the `currency` params to `_xendit_create_charge` following [my comment](https://github.com/odoo/odoo/pull/158445#discussion_r1539847023).

As I asked you in my previous comment, please write a good commit message and document your changes in the `readme.md` file of `payment_xendit`. Also, don't forget to answer all unresolved comments and leave them unresolved if you answered so I can see it. (you can simply resolve them if you decide to apply the changes :smile:  ).

Have a nice day!

nni-odoo commented:
Thanks again @chevalierv ! Missed the commit message & README edit message my bad.
I've changed the original commit message and added some more things on the README. Let me know if that's enough.
Other than that,  I've also adapted the implementation according to your suggestions! 
Have  a good day!

chevalierv reviewed (changes requested):
Hello @nni-odoo 

Here is my review :smiley: 

Don't forget to squash your commits and apply or answer all comments. For example, see https://github.com/odoo/odoo/pull/158445#discussion_r1539756202.

Regarding your commit message, please focus more on the why than on the what. Here is an example :
```
Currently, in the APAC region, Odoo doesn't support tokenization on any available payment provider in this area. 

This commit adds tokenization support for the `Card` payment method through Xendit. As Xendit does not offer an easy redirect implementation for tokenization, `Card` payments will now be direct. This does not affect other payment methods supported by Xendit which are still using the redirect implementation.
```

Have a nice day!

chevalierv commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,212 @@
+
+        Object.keys(formInputs).forEach(el => formInputs[el].setCustomValidity(""))
+        if(!Xendit.card.validateCardNumber(details.card_number)){
+            formInputs['card'].setCustomValidity("Invalid Input")
```
Also, please ensure that those messages will be translated, as any other message in your PR.
```suggestion
            formInputs['card'].setCustomValidity(_t("Invalid Card Number"))
```

nni-odoo commented on addons/payment_authorize/models/payment_transaction.py:
```diff
@@ -227,7 +227,7 @@ def _process_notification_data(self, notification_data):
             if status_type in ('auth_capture', 'prior_auth_capture'):
                 self._set_done()
-                if self.tokenize and not self.token_id:
+                if self.tokenize:
```
oh right my bad, I edited the wrong file

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,212 @@
+            formInputs['card'].setCustomValidity("Invalid Input")
+        }
+        if(!Xendit.card.validateExpiry(details.card_exp_month, details.card_exp_year)){
+            formInputs['month'].setCustomValidity("Invalid Input")
+            formInputs['year'].setCustomValidity("Invalid Input")
```
@chevalierv for this one, I think I'll do `_t("Invalid Date")` for both form inputs because suppose date is (10/2022), the invalid one is year and it would be pretty misleading to say month is invalid. Let me know if you have a better sugggestion

chevalierv replied:
LGTM

nni-odoo commented:
Hello @chevalierv , thank you again for the reviews!
I've made change to commit message and put more details on the README. 
About squashing commits, not sure if you want me to squash it now or later on towards the end of review when code is about to be merged? Let me know if there's anything else to fix

Nice day to you too!

chevalierv commented:
Hello @nni-odoo 

I'd like you to squash your commits now, there is no need to wait :)

Also, please answer all unresolved comments (I see that https://github.com/odoo/odoo/pull/158445#discussion_r1539756202 is still unresolved)

nni-odoo commented:
Hello @chevalierv squashed all the commits and resolved all the unanswered comments!

chevalierv commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,207 @@
+     *
+     * @private
+     * @param {number} paymentOptionId - The id of the selected payment option.
+     * @param {string} paymentMethodCode - The code of the selected payment method, if any.
```
```suggestion
```

chevalierv commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,207 @@
+     *
+     * @private
+     * @param {number} paymentOptionId - The id of the selected payment option.
+     * @param {string} paymentMethodCode  - The code of the selected payment method, if any.
```
```suggestion
```

chevalierv reviewed (approved):
Still two minors comments :) Please fix it straight away (I mean don't wait for the final review)

Otherwise it LGTM :+1: 

Thanks for your collaboration :smiley:

chevalierv commented on addons/payment_xendit/models/payment_transaction.py:
```diff
@@ -27,13 +27,13 @@ def _get_specific_rendering_values(self, processing_values):
         """
         res = super()._get_specific_rendering_values(processing_values)
-        if self.provider_code != 'xendit':
+        if self.provider_code != 'xendit' and self.payment_method != 'card':
```
Your runbot is red :eyes: 
```suggestion
        if self.provider_code != 'xendit' and self.payment_method_code != 'card':
```

nni-odoo commented:
@chevalierv fixed the issues and I've re-tested the whole flow again, seems to be smooth from my side.

Thank you again for your help on this!!

chevalierv commented on addons/payment_xendit/models/payment_transaction.py:
```diff
@@ -122,7 +122,7 @@ def _process_notification_data(self, notification_data):
 
         super()._process_notification_data(notification_data)
-        if self.provider_code != 'xendit':
+        if self.provider_code != 'xendit' and self.payment_method != 'card':
```
runbot is red :eyes: 
Why did you add this? We need to process all transactions regardless of the payment method.
```suggestion
        if self.provider_code != 'xendit':
```

nni-odoo commented:
@chevalierv my bad on that one. I pushed again the fixes but it seems there's still red on runbot. When I look into it, the error seems to not be correlated to the changes introduced?

chevalierv commented:
Hello @nni-odoo 

On this specific PR, we expected to see a red runbot for CI/template and CI/security. We'll take care of it we'll do the final review. 

For the ci/runbot, I just checked and IMO, it's not related to your changes. You could try to rebase your branch to see if the issues has been fixed.

nni-odoo commented:
@chevalierv rebased and runbot is green now! thanks!

AntoineVDV commented on addons/payment_xendit/controllers/main.py:
```diff
@@ -58,3 +58,12 @@ def _verify_notification_token(self, received_token, tx_sudo):
+        :param str reference: The reference of the transation
+        """
+        tx_sudo = request.env['payment.transaction'].sudo().search([('reference', '=', reference)])
+        tx_sudo._xendit_create_charge(token_id)
```
I'm a bit concerned by the security aspect of this call. Basically, if an attacker finds someone else's token ID, they can use it to pay for their own transaction that was created beforehand.

nni-odoo replied:
`_xendit_create_charge` calls `_xendit_make_request`  which will still require the `xendit_secret_key` field when calling the API . So I think that attacker would only be successful when he knows both the token and also the secret key. 
If the secret key is already exposed, I think attacker would pretty much be able to do anything :laughing:

AntoineVDV replied:
They can call the route with an arbitrary reference and token ID, so they would actually only need to know the latter. I will check with the security team.

AntoineVDV replied:
@odoo/rd-security could we have your input here?

nni-odoo replied:
@xmo-odoo thank you for your input here! Indeed, Xendit allows a single-use token as well (make the is_multiple_use: False)

In this case, I've made the change such that single/multi-use is conditional on whether user wants to save the payment method. That's what I'm getting from your feedback, but  please do correct me if I'm wrong or if there's still anything I miss!
cc @AntoineVDV

AntoineVDV commented on addons/payment_xendit/controllers/main.py:
```diff
@@ -58,3 +58,12 @@ def _verify_notification_token(self, received_token, tx_sudo):
             raise Forbidden()
+
+    @http.route('/payment/xendit/payment', type='json', auth='public')
+    def xendit_payment(self, reference, partner_id, token_id):
```
`partner_id` is never used.

AntoineVDV commented on addons/payment_xendit/controllers/main.py:
```diff
@@ -58,3 +58,12 @@ def _verify_notification_token(self, received_token, tx_sudo):
+        :param str reference: The reference of the transation
+        """
+        tx_sudo = request.env['payment.transaction'].sudo().search([('reference', '=', reference)])
+        tx_sudo._xendit_create_charge(token_id)
```
Shouldn't you rather create the token in Odoo and set Xendit's token_id as `provider_ref`, then call `_send_payment_request`?

nni-odoo replied:
this is true. The reason I do this is, when I was developing it, the token created turns out to be invalid (because it was not multiple-use), which in this case results to unsuccessful payment, however the token is still created somehow.
So I'm doing this for safety reason to make sure that payment is successful first before storing the token in Odoo. Let me know what you think and if we should still proceed with changing the logic here

nni-odoo replied:
that sounds like a good idea, will do that thanks!

nni-odoo replied:
Hello @AntoineVDV for this one, I tried implementing it and have some concerns with the approach mentioned:

* we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one. 
* The possible option that I see is overriding the `_initiatePaymentFlow()`, create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details. 
* Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.
* If we override rom the `_processDirectFlow`, the transactions is already created, not sure if it's a good idea to attach the token_id to transaction from this point on?

For these points, I believe it's best to keep the current implementation as is. Let me know if you think there's something better to do here. I checked all other payment methods now including SEPA and I don't think this way is currently implemented anymore...

AntoineVDV replied:
> - Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.

That was my main concern and reason not to go this road.

Let's keep the current solution then ;)

> - we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one.
> - The possible option that I see is overriding the _initiatePaymentFlow(), create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details.
> - If we override rom the _processDirectFlow, the transaction is already created, not sure if it's a good idea to attach the token_id to transaction from this point on? If we do the _initiatePaymentFlow , this means we will create a new transaction again.

We have equivalent methods in 17.0 that could have met our requirements but no need to bother anymore if we create the token after payment.

nni-odoo replied:
Noted with thanks! will resolve the convo in that case

AntoineVDV commented on addons/payment_xendit/models/payment_transaction.py:
```diff
@@ -140,11 +140,78 @@ def _process_notification_data(self, notification_data):
+    def _xendit_tokenize_from_notification_data(self, notification_data):
+        """ Create a new token based on the notification data.
+
+        :param dict notification_data: The notification data built with Xendit objects.
+                                       See `_process_notification_data`.
```
What do you mean by "...Xendit objects. See _process_notification_data"?

AntoineVDV reviewed (changes requested):
Hi, sorry for the late review :innocent: 

I applied some changes directly in a fixup commit (after rebasing on master) to save time with the review, but I still had some remarks/questions that I posted here below.

Also, I'm getting this error when trying to process a validation transaction (through /my/payment_method)
![image](https://github.com/odoo/odoo/assets/15896813/71e7d702-1cac-47f1-b7c0-d5eb516c7b89)
I might have broken something without noticing... could you check? :pray: 

Finally, could you make sure that the test coverage is complete?

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,205 @@
+            should_authenticate: false,
+        }
+
+        Xendit.card.createToken(
```
It will not be allowed, I believe an error would be thrown right away while trying to create token (which I believe is already handled currently), so I suppose no adjustmends needed here?

AntoineVDV replied:
No need, indeed; I just wanted confirmation ;)

nni-odoo commented on addons/payment_xendit/static/src/js/payment_form.js:
```diff
@@ -0,0 +1,205 @@
+            formInputs['month'].setCustomValidity(_t("Invalid Date"))
+            formInputs['year'].setCustomValidity(_t("Invalid Date"))
+        }
+        if(!Xendit.card.validateCvn(details.card_cvn)){
```
I'll make the input required in this case

nni-odoo commented:
@AntoineVDV made the changes on PR per your review , as for the error you've screenshoted, I think I encountered before based on the error, and I believe it should be related to the credit card payment configuration? (try this URL after logging in: https://dashboard.xendit.co/settings/payment-methods/cards-configuration)
![image](https://github.com/odoo/odoo/assets/82074749/11defe70-2e7b-437d-9bf1-68276884fee2)

nni-odoo commented:
@AntoineVDV noted on that, I've made the PR for docu update here: https://github.com/odoo/documentation/pull/9394

Thank you again!

AntoineVDV commented:
Hello @xmo-odoo could you please check https://github.com/odoo/odoo/pull/158445#discussion_r1604595805? :pray:

nni-odoo commented:
@xmo-odoo 
for `should_authenticate`, it's actually for convenience purposes so that it won't ask for verification, I'll look into the flow if we let `should_authenticate` to true.

To my understanding from reading the code, isn't the `paymentContext.tokenizationRequested` related to whether or not user checks the "Save My Payment Details" when paying with credit card, which then is passed when creating transaction (field: tokenize). This is where I put the condition of `self.tokenize`, which will create the token in Odoo and save it to the transaction too. So with current implementation, single-use tokens won't be saved but multi-use tokens will.

Hope this addresses your concern and let me know if I missed anything here,
thank you again for reviewing!

nni-odoo commented:
@xmo-odoo @AntoineVDV To follow up on earlier statement regarding the `should_authenticate=False`. If it's true, it will opt for user verification as such 
![image](https://github.com/odoo/odoo/assets/82074749/ab62e91f-1859-44d9-accf-341e2ec5e0a0)

I'm wondering if we've implemented something similar before? If so, any reference I could use?
On the other hand if not, is it something you think is worth exploring on? 

Let me know if you have further questions!

nni-odoo commented:
@AntoineVDV adding little context to the latest commit

since we're adding authentication flow, I'm taking reference from xendit themselves:
- https://docs.xendit.co/credit-cards/integrations/tokenization (reference)
- https://js.xendit.co/test_tokenize.html (sample form)

I'm adding `auth_ui.js` and `auth_service.js` that follows how `block_ui` is applied. Since we need to still have the BlockUI (to prevent user choosing other payment methods), then we'll have to create new component "above" BlockUI. I came up with adding z-index but z-index will only work with blocks at the same level (ref: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context). And that's why I have to put it into the `main_component`. 

Let me know if you have other suggestions that could simplify this, thanks!

AntoineVDV reviewed (approved):
Hello @nni-odoo, I think we're good to go! :rocket: 

TBH, I have doubts about this new auth_ui/service thing because, as you said, it's a bit "hacky-hacky". There might be a simpler way to achieve the same result, but since it's only frontend assets (it can be changed easily), does work, and is not worth spending too much time on, let's keep it as-is ;)

AntoineVDV commented:
@xmo-odoo Could you please add the ignore ci/security when you consider all security concerns have been answered?

AntoineVDV commented:
@robodoo r+
@robodoo override=ci/security
Reason: `window.location`: Redirection to a fixed, hard-coded, relative URL.
Link to ci/security build (master): https://runbot.odoo.com/runbot/build/64109175

robodoo commented:
@AntoineVDV you may want to rebuild or fix this PR as it has failed CI.
